### PR TITLE
Introduce packaging tests for Docker

### LINKED ISSUE #37617: Add docker packaging tests
Now that we have our docker distribution in the elasticsearch repo, we should add packaging tests (of some sort) that ensure the bits inside the docker distribution work as intended. This would help with changes like #37243 to allow confidence in the change.

### PR DESCRIPTION
**Pull request by pugnascotia**
Closes #37617. Add packaging tests for our Docker images, similar to what we have for RPMs or Debian packages. This works by running a container and probing it e.g. via `docker exec`. Test can also be run in Vagrant, by exporting the Docker images to disk and loading them again in VMs.

Notes:

   * Docker is installed via `Vagrantfile` in a selection of boxes. We could change the boxes built by Infra if we felt it was worthwhile, I wasn't sure.
   * The Docker images are only tested with the bundled JDK. It seemed counter to the spirit of using containers to run one using a bundled JDK, though there's nothing stopped a user doing that.

### CODE REVIEW

**Review by dliappis — CHANGES_REQUESTED** [review]
Great work this is really needed after integrating the Docker parts from the old separate repo!

I took an initial pass and left some comments.

**Review by alpar-t — COMMENTED** [review]
Thanks for taking this on! 
Unfortunately we don't have Pr checks for running these directly on jenkins workers, so it would be nice to test this before merging to make sure the CI jobs don't break.

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/packaging

**Comment by pugnascotia** [comment]
Mysterious - for some reason, the `centos-6` Vagrant box is not only trying to install Docker when it shouldn't be, it's trying to do with `apt-get`! 🤷‍♂

File: Vagrantfile
Comment (dliappis) [inline Vagrantfile]: Is `run: 'always'` here on purpose? Do we intend to attempt to install docker on every `vagrant up/reload` operation? Details [here](https://www.vagrantup.com/docs/provisioning/basic_usage.html#run-once-always-or-never).

File: Vagrantfile
Comment (dliappis) [inline Vagrantfile]: Same comment here as before for `run: 'always'`.

File: Vagrantfile
Comment (dliappis) [inline Vagrantfile]: Same comment here as before for `run: 'always'`.

File: Vagrantfile
Comment (dliappis) [inline Vagrantfile]: This should be moved before `end` in line 91; having it here causes `deb_docker` to be invoked for every machine.

**Review by dliappis — CHANGES_REQUESTED** [review]
Left a few more contents related to the Vagrantfile.

**Comment by pugnascotia** [comment]
Thanks @dliappis - facepalms aplenty. 🤦‍♂

**Comment by pugnascotia** [comment]
@elasticmachine update branch

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/packaging

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/packaging as there was a transient error.

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/packaging as Jenkins restarted.

**Comment by pugnascotia** [comment]
CI is green if someone would like to give me a LGTM?

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java
Comment (atorok) [inline buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: @pugnascotia the download distribution plugin is meant to be used outside of our build,
@rjernst is working on adding support to download some of the artifacts that are not already downloadable.  I think in the case of docker, this could be set up differently to make that work easier.
For example, `ensureImageIsLoaded` from the test should be part of this plugin, since docker is distributed trough the registry, so there will be nothing to do here for the released versions. 

Alternatively @rjernst,  it might be more straight forward and easier to maintain to split the download plugin into two. One that deals exclusively with versions that are an actual download, and restrict that to some distribution types and one that deals with unreleased versions, with the lather being applied only as part of our build. 

I don't necessarily see this as reason  not to merge this PR.

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java
Comment (atorok) [inline qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java]: Could we implement this similarly to other distributions by checking the cluster health endpoint ? 
It's counter intuitive to have a failure in a packaging test when changing a log line.

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java
Comment (atorok) [inline qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java]: does this still need to be looked at ?

Comment (pugnascotia) [reply qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java]: Yes - I planned to look at this separately once the PR is merged.

**Review by dliappis — APPROVED** [review]

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java
Comment (pugnascotia) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: Yes, I feel that we should merge this and come back to some of these issues. At least we'll have something build upon.

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java
Comment (pugnascotia) [reply qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java]: Thanks, I've done as you suggested.

**Review by alpar-t — APPROVED** [review]
LGTM 

We still have to make sure this works with the new workers.
If #46924 is merged, PR checks for it are coming soon, if this happens to be merged before, I'll re test to make sure before merging mine.

**Comment by pugnascotia** [comment]
@atorok is your PR close to merging? I don't mind waiting if it is.

**Comment by atorok** [comment]
> @atorok is your PR close to merging? I don't mind waiting if it is.

Yes it is.

**Comment by pugnascotia** [comment]
@elasticmachine update branch

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java]: This should be handled outside of addDistro, otherwise we are adding the same distribution twice? ie around line 325 where this method is called

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java]: This shouldn't be necessary, see my previous comment about where we add the distros. We should never add a docker distro with bundledJdk=false

File: distribution/docker/build.gradle
Comment (rjernst) [inline distribution/docker/build.gradle]: Please use static types. It makes the code easier to reason about, eg String here.

File: qa/os/src/test/java/org/elasticsearch/packaging/test/DockerTests.java
Comment (rjernst) [inline qa/os/src/test/java/org/elasticsearch/packaging/test/DockerTests.java]: This doesn't really match with the pattern we've been working towards. We want test classes for categories of checks, like keystore related (which @williamrandolph has been splitting out in a separate PR). Which distribution is being used should not matter to the test, only that we know how to do primitive operations for the given distribution (eg start, stop, run cli tool, etc)

File: qa/os/src/test/java/org/elasticsearch/packaging/test/DockerTests.java
Comment (rjernst) [inline qa/os/src/test/java/org/elasticsearch/packaging/test/DockerTests.java]: We definitely don't want to have copies of every test in archive tests. Most of that class needs to be broken up. I wonder if in the interum ArchiveTests could be adapted to run when docker type is used?

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Distribution.java
Comment (rjernst) [inline qa/os/src/test/java/org/elasticsearch/packaging/util/Distribution.java]: Why is this needed? It doesn't seem any clearer than `packaging == Packaging.DOCKER`

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java
Comment (rjernst) [inline qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java]: I don't think we should copy everything that is in Archives. This can (and will) get out of sync. I think we need to think more about how to decouple the checks so they can run agnostically.

**Review by rjernst — CHANGES_REQUESTED** [review]
Thanks for the work here! Better testing of our docker images is greatly needed.

After reading through this, it appears we are running docker within each VM. At first I was confused by this, but I guess this matches up with docker being another distribution, and we want to test it on each platform (similar to how we test java on each platform).

One general thing I noticed though is we seem to have duplicated a lot of the assertions that exist within the archive tests. This would mean making any change to the archive structure would require changing multiple tests. Could we reuse these assertions?

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Distribution.java
Comment (pugnascotia) [reply qa/os/src/test/java/org/elasticsearch/packaging/util/Distribution.java]: To be honest I was copying the pattern of the existing code. e.g. `isArchive()` is only called in one place 🤷‍♂. I can remove it.

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java
Comment (pugnascotia) [reply qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java]: How about I do that in a follow-up PR? This one has been waiting to merge for a while. I can then take the time to refactor the packaging tests more generally.

**Comment by pugnascotia** [comment]
I'm going to merge this PR as-is, and once #44775 is merged, I'll refactor the packaging tests to reduce duplication. The infrastructure here will unblock @dliappis and another PR of mine.

**Comment by atorok** [comment]
@pugnascotia this broke the new packaging tests: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob+packaging-tests-unix-next/os=ubuntu-16.04-packaging/13/console

**Comment by atorok** [comment]
@pugnascotia sorry I missed this during review, we really want to back-port testing and build related changes like this, especially anything that changes the build code needs to be back-ported so the code doesn't diverge making it easy to port improvement to older branches that are still built in CI.

**Comment by pugnascotia** [comment]
No problem, happy to backport.

Re: the CI failure, it looks like the `docker build` step failed, which I'm pretty sure is transient and not related to my changes. `yum` is complaining about its mirrors.

**Comment by atorok** [comment]
I added more details to #47639 there's  more to it, and some of the tests fail too

**Pull request merged**
